### DISCUSSION TITLE
What is the best way to migrate less files from v4 to V5

### DISCUSSION THREAD

**Discussion by AmrAhmedA** [ant-design/ant-design#40317] [Q&A] [answered]
I'm currently looking for document dedicated for the less migration from v4 to v5. 
I understand that you have discarded less entirely in your v5. 
One of the biggest challenges that you guys didn't yet address in details and how to handle them efficiently is migration from less files to antd [cssinjs](https://github.com/ant-design/cssinjs).  

If there is huge code base being dependent on less files with the global variables in it. and out of sudden you abandoned less. what is the best way to handle this?
I assume that migration from less to cssinjs is big change and would require decent amount of time. there should be migration manual dedicated entirely for less to cssinjs.

**Comment by arvinxx** [comment, marked as answer]
Edit on 2023.08.29
We have release the antd-style ,and please check our documents here: https://ant-design.github.io/antd-style/

-----

we are working on this. You can find some example in this link. https://ant-design.github.io/antd-style/guide/migrate-less-codemod
But there's Chinese only by now.(You can use translator to read it).And this application level cssinjs solution will release on 2023.02 later or March maybe.

**Reply by stanbar** [reply]
The link does not work. Any update on this?

**Reply by arvinxx** [reply]
Here it is. https://ant-design-antd-style-preview-pr-19.surge.sh/guide/migrate-less-application

**Reply by Histamina** [reply]
> we are working on this. You can find some example in this link. https://antd-style-git-alpha-arvinxx.vercel.app/guide/migrate-from-less But there's Chinese only by now.(You can use translator to read it).And this application level cssinjs solution will release on 2023.02 later or March maybe.

Hi. Is there any news about the solution release that you mentioned? Thanks.

**Reply by arvinxx** [reply]
> 
> Hi. Is there any news about the solution release that you mentioned? Thanks.

You can try now on beta version


npm i antd-style@beta

We will release it this month

**Reply by dazaart** [reply]
@arvinxx can you maybe explain it, is it a version of antd that didnt drop less?

**Reply by dazaart** [reply]
I installed the beta version, but it still throws errors with global less files imports.

**Reply by abenhamdine** [reply]
@daryazata  I think you should read https://ant-design-antd-style-preview-pr-19.surge.sh/guide/migrate-less-application

**Reply by Histamina** [reply]
They should put a proper documentation for this. It's unbeliveable that antd uses less until V4 inclusive, change to CSS in JS (I understand why) in V5 and then expects that we sudenly change all our projects structure with some lines of migration code context. [https://ant.design/docs/react/migration-v5#less-migration](url)
In order to achieve this (use of less in v5) and mantain the sustentability of Ant Desing in our work enviroments, could you invest some time for a controled usability step by step documentation?

**Reply by abenhamdine** [reply]
> then expects that we sudenly change all our projects structure with some lines of migration code context. [https://ant.design/docs/react/migration-v5#less-migration](https://github.com/ant-design/ant-design/discussions/url)

I don't think they expect anything, Antdesign is OSS and free, and moreover you don't have to migrate on V5, antd 4 is still maintained

**Reply by dazaart** [reply]
@abenhamdine well Im not sure it should be the way how even open source libraries develop, that they suddenly leave many users behind by breaking changes (and dont even announce it as breaking change). Ofcourse you _can_ do anything, but its not very thoughtfull. 
[https://ant-design-antd-style-preview-pr-19.surge.sh/guide/migrate-less-application](https://ant-design-antd-style-preview-pr-19.surge.sh/guide/migrate-less-application) how can this help with global styles(theme)?

**Reply by stefanonardo** [reply]
did the migration docs improve since this post? The URL posted by @arvinxx does not work anymore

**Reply by abenhamdine** [reply]
yes, the official docs are here : https://ant.design/docs/react/migration-v5#less-migration  
But I strongly advise to wait for late august or early september, all design component tokens will be available then (it's not the case yet) and it will be easier to customize style of some components

**Reply by atanasov-deyan** [reply]
Hi @arvinxx,  I'm stuck in trying to figure out a feasible way of migrating to antd v5 and came across this thread. Sadly, trying to access link above: https://antd-style-git-alpha-arvinxx.vercel.app/guide/migrate-from-less 

throws a 404 error as below:
![image](https://github.com/ant-design/ant-design/assets/64613743/3cd8ee94-7315-4ca8-8fdf-b5e7518056a9)

any way we can revive it?

Thanks!

**Reply by abenhamdine** [reply]
you should await a bit more IMHO
last remaining missing design tokens and classNames are in feature branch and should be in next weekly release

**Reply by arvinxx** [reply]
@atanasov-deyan Here: https://ant-design.github.io/antd-style/guide/migrate-less-codemod

**Reply by abenhamdine** [reply]
> last remaining missing design tokens and classNames are in feature branch and should be in next weekly release

actually it should be in next monthly version 5.9.0, in a few days

**Reply by dagadbm** [reply]
hey all. I am also trying to migrate to v5 but the less tokens present here: 

```
const mapV5Token = defaultAlgorithm(defaultSeed);
const v5Vars = convertLegacyToken(mapV5Token);
const mapV4Token = theme.getDesignToken(defaultTheme);
const v4Vars = convertLegacyToken(mapV4Token);
```

do not contain the `prefix-cls` less variables.

what can I do in this case?

I looked at that codemode but it forces me to install yet another thing and I just want ant less to still work as before

**Reply by dagadbm** [reply]
So I went to old ant v4 code and found the prefix-cls names and just replaced them directly on my code. just in case someone stumbles upon it. easiest solution I could think of.
